libevmasm/Inliner.cpp

```diff
@@ -0,0 +1,216 @@
+	)
+	{
+		exitJump.setJumpType(AssemblyItem::JumpType::Ordinary);
+
+		// Accumulate size of the inline candidate block in bytes (without the return jump).
+		uint64_t functionBodySize = codeSize(ranges::views::drop_last(_block.items, 1));
+
+		// Use the number of push tags as approximation of the average number of calls to the function per run.
+		uint64_t numberOfCalls = _block.pushTagCount;
+		// Also use the number of push tags as approximation of the number of call sites to the function.
+		uint64_t numberOfCallSites = _block.pushTagCount;
+
+		static AssemblyItems const uninlinedCallSitePattern = {
+			AssemblyItem{PushTag},
+			AssemblyItem{PushTag},
+			AssemblyItem{Instruction::JUMP},
+			AssemblyItem{Tag}
+		};
+		static AssemblyItems const uninlinedJumpSitePattern = {
+			AssemblyItem{Tag},
+			// Actual function body of size functionBodySize. Handled separately below.
+			AssemblyItem{Instruction::JUMP}
+		};
+
+		// Both the call site and jump site pattern is executed for each call.
+		// Since the function body has to be executed equally often both with and without inlining,
+		// it can be ignored.
+		bigint uninlinedExecutionCost = numberOfCalls * (
+			executionCost(uninlinedCallSitePattern, m_evmVersion) +
+			executionCost(uninlinedJumpSitePattern, m_evmVersion)
+		);
+		// Each call site deposits the call site pattern, whereas the jump site pattern and the function itself are deposited once.
+		bigint uninlinedDepositCost = GasMeter::dataGas(
+			numberOfCallSites * codeSize(uninlinedCallSitePattern) +
+			codeSize(uninlinedJumpSitePattern) +
+			functionBodySize,
+			m_isCreation,
+			m_evmVersion
+		);
+		// When inlining the execution cost beyond the actual function execution is zero,
+		// but for each call site a copy of the function is deposited.
+		bigint inlinedDepositCost = GasMeter::dataGas(numberOfCallSites * functionBodySize, m_isCreation, m_evmVersion);
+
+		// If the estimated runtime cost over the lifetime of the contract plus the deposit cost in the uninlined case
+		// exceed the inlined deposit costs, it is beneficial to inline.
+		if (bigint(m_runs) * uninlinedExecutionCost + uninlinedDepositCost > inlinedDepositCost)
+			return exitJump;
+	}
+
+	return nullopt;
+}
+
+
+void Inliner::optimise()
+{
+	std::map<u256, InlinableBlock> inlinableBlocks = determineInlinableBlocks(m_items);
+
+	if (inlinableBlocks.empty())
+		return;
+
+	AssemblyItems newItems;
+	for (auto it = m_items.begin(); it != m_items.end(); ++it)
+	{
+		AssemblyItem const& item = *it;
+		if (next(it) != m_items.end())
+		{
+			AssemblyItem const& nextItem = *next(it);
+			if (item.type() == PushTag && nextItem == Instruction::JUMP)
+				if (auto* inlinableBlock = util::valueOrNullptr(inlinableBlocks, item.data()))
+					if (auto exitJump = shouldInline(item.data(), nextItem, *inlinableBlock))
+					{
+						newItems += ranges::views::drop_last(inlinableBlock->items, 1);
+						newItems.emplace_back(*exitJump);
```

Comment:
Wouldn't it be better to just return the jump type?
```suggestion
					if (auto exitJumpType = shouldInline(item.data(), nextItem, *inlinableBlock))
					{
						newItems += inlinableBlock->items;
						newItems.back().setJumpTypE(exitJump);
```

Reply:
Hm... unfortunately that's again only defined inline in ``AssemblyItem``, so I can't forward-declare it... maybe just properly including ``AssemblyItem`` there isn't too bad in the end...